src/register/sireg.rs

```diff
@@ -0,0 +1,162 @@
+//! sireg register
+//!
+//! The `sireg` CSR is defined in "The RISC-V Advanced Interrupt
+//! Architecture" Version 0.3.2-draft
+//!
+//! Before accessing `sireg` using the functions in this module,
+//! software must select the indirect register using the `siselect` CSR.
```

Comment:
Why give the flexibility to software to make mistakes? I understand rest of this crate is organized such that it gives individual CSR access to caller. But, siselect/sireg are special. These are window registers that provide access to a large set of registers. What do you think about only providing access to caller to perform:

```
sireg_read_eidelivery() -> usize
sireg_write_eidelivery(usize)

sireg_read_eithreshold() -> usize
sireg_write_eithreshold(usize)

sireg_read_eip(usize) -> EipStatus
sireg_write_eip(usize, EipStatus)

sireg_read_eie(usize) -> EieStatus
sireg_write_eie(usize, EieStatus)
```

For eip and eie we don't even need to expose the actual reg # because the caller will have to calculate which bit in which eipX / eieX register corresponds to an interrupt #. Thus, instead these functions can take interrupt # as input and do the necessary calculation for determining what CSR to read/write.

Reply:
Yeah, I think you're right. I was trying to find the right balance in the API between making the code simple enough and matching the style of the rest of the crate; versus making the siselect/sireg interface easier (and safer) to use. I've rewritten this interface to make sireg the only interface one needs to use to access the registers (and improved the interrupt enables and interrupt pending bits by making the sireg interface aware of how the bits are packed into registers, to make the interface easier to use).

Reply:
As I was reading your code, I realized that we could have solved this in another way, probably slightly more cleaner w.r.t. internal implementation and from caller standpoint.

1. Make sireg and siselect both private i.e. callers never see the regs directly. These are only for internal usage.
2. Create public facing CSRs for eidelivery, eithreshold, eip, eie and provide implementations that make use of sireg and siselect to do their work. We can also provide some riscv-specific register macros to provide the boilerplate code for `read_windowed_csr`/`write_windowed_csr`.

Anyways, I think what you have here is easy to use as well. From a caller standpoint, the only difference would have been: Instead of using `register::sireg_eidelivery::read()`, they would use `register::eidelivery::read()`.

When we move to tock registers, maybe we can keep that in mind.

Reply:
As I was reading your code, I realized that we could have solved this in another way, probably slightly more cleaner w.r.t. internal implementation and from caller standpoint.

Make sireg and siselect both private i.e. callers never see the regs directly. These are only for internal usage.
Create public facing CSRs for eidelivery, eithreshold, eip, eie and provide implementations that make use of sireg and siselect to do their work. We can also provide some riscv-specific register macros to provide the boilerplate code for read_windowed_csr/write_windowed_csr.
Anyways, I think what you have here is easy to use as well. From a caller standpoint, the only difference would have been: Instead of using register::sireg_eidelivery::read(), they would use register::eidelivery::read().

When we move to tock registers, maybe we can keep that in mind.